inception-search-core/src/main/java/de/tudarmstadt/ukp/inception/search/SearchServiceImpl.java

```diff
@@ -179,48 +193,102 @@ public void beforeDocumentRemove(BeforeDocumentRemovedEvent aEvent) throws Excep
         }
     }
 
-    @TransactionalEventListener(fallbackExecution = true)
-    @Transactional
-    public void afterAnnotationUpdate(AfterAnnotationUpdateEvent aEvent) throws Exception
+    private boolean canAddDocumentToIndex(Index aIndex)
     {
-        AnnotationDocument document = aEvent.getDocument();
-
-        Project project = document.getProject();
-        
-        // Retrieve index entry for the project
-        Index index = getIndexFromMemory(project);
-
-        if (!index.getPhysicalIndex().isCreated()) {
+        if (!aIndex.getPhysicalIndex().isCreated()) {
             // Physical index does not exist. 
+            log.debug("Physical index not created. Set it to invalid and enqueue reindex task.");
             
             // Set the invalid flag
-            index.setInvalid(true);
-            updateIndex(index);
+            aIndex.setInvalid(true);
+            updateIndex(aIndex);
 
             // Schedule new reindex process
-            indexScheduler.enqueueReindexTask(project);
-        }
-        else {
+            indexScheduler.enqueueReindexTask(aIndex.getProject());
+            
+            return false;
+        } else {
             // Physical index already exists
+            log.debug("Physical index already created.");
 
-            if (!index.getPhysicalIndex().isOpen()) {
+            if (!aIndex.getPhysicalIndex().isOpen()) {
                 // Physical index is not open. Open it.
-                index.getPhysicalIndex().openPhysicalIndex();
+                log.debug("Physical index not open. Open it");
+
+                aIndex.getPhysicalIndex().openPhysicalIndex();
             }
 
-            // Remove annotation document from index.
-            index.getPhysicalIndex().deindexDocument(document);
+            return false;
+        }
+    }
+    
+    @Override
+    public void indexDocument(SourceDocument aSourceDocument, JCas aJCas)
+    {
+        // Retrieve index entry for the project
+        Index index = getIndexFromMemory(aSourceDocument.getProject());
+
+        if (canAddDocumentToIndex(index)) {
```

Comment:
I would invert the condition and return early, then you can remove one nested level.

Reply:
Done.